ec2_scaling_policy boto3 rewrite for step scaling

willthames opened the pull request:
##### SUMMARY
Convert ec2_scaling_policy to boto3 and implement step scaling

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ec2_scaling_policy

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel e612d0be12) last updated 2017/07/05 16:42:21 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```


##### ADDITIONAL INFORMATION
Tests require IAM privileges to create LaunchConfigs, AutoScalingGroups and Scaling Polices

willthames commented:
WIP until #26233 is merged so that the IAM policies can be put in place in a modular fashion

gundalow commented:
As agreed in 2017-12-07 Core meeting, we've done bulk pep8 fixes across all modules in `devel`.

This was done in https://github.com/ansible/ansible/pull/33682 and https://github.com/ansible/ansible/pull/33690 which have been merged into `devel`. because of this you will need to rebase your PR. Thanks in advance.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_scaling_policy.py:
```diff
@@ -1,10 +1,18 @@
-from __future__ import absolute_import, division, print_function
-__metaclass__ = type
-
+# This file is part of Ansible
```
Isn't the current copyright right?

willthames replied:
Umm, it does look that way doesn't it. Again, an old change before the copyright standard was locked down, at least in my head!

willthames replied:
Ah, I know what happened - the copyright and `from __future__` stuff got changed well after this PR was made in  9c81257c

s-hertel commented on test/integration/targets/ec2_scaling_policy/aliases:
```diff
@@ -0,0 +1,2 @@
+cloud/aws
+posix/ci/cloud/aws
```
This alias needs to be updated to run in CI

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_scaling_policy.py:
```diff
@@ -31,155 +41,353 @@
+    try:
+        policies = connection.describe_policies(AutoScalingGroupName=asg_name,
+                                                PolicyNames=[policy_name])['ScalingPolicies']
+    except botocore.exceptions.ClientError as e:
```
BotoCoreError should be handled for the API calls as well.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_scaling_policy.py:
```diff
@@ -31,155 +41,353 @@
-        module.fail_json(msg=str(e))
+        connection = boto3_conn(module, conn_type='client', resource='autoscaling',
+                                region=region, endpoint=ec2_url, **aws_connect_params)
+    except (botocore.exceptions.NoCredentialsError, botocore.exceptions.ProfileNotFound) as e:
```
The connection doesn't need exception handling in the module. You could also update this to use AnsibleAWSModule and connection = module.client('autoscaling') if you wanted.

willthames replied:
Yeah, AnsibleAWSModule didn't exist when I made this PR :)

s-hertel reviewed (commented):
A few trivial things, but looks great besides that.

Commit:
Enable step scaling for ec2_scaling_policy

Convert ec2_scaling_policy module to boto3 to allow step scaling
Provide parameters to allow step scaling

Commit:
Add test suite for ec2_scaling_policy

mkrizek commented:
Thank you very much for your interest in Ansible. This plugin is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.aws.

Please re-submit this issue in the above repository.

If you have further questions please stop by IRC or the mailing list:

* IRC: #ansible on irc.freenode.net
* mailing list: https://groups.google.com/forum/#!forum/ansible-project
